proxy/src/control/cache.rs

```diff
@@ -1,30 +1,39 @@
-use std;
-use std::collections::HashSet;
+use indexmap::{map, IndexMap};
+use std::borrow::Borrow;
+use std::hash::Hash;
+use std::iter::IntoIterator;
 use std::mem;
 
-/// A cache that supports incremental updates with lazy resetting on
-/// invalidation.
+/// A key-value cache that supports incremental updates with lazy resetting
+/// on invalidation.
 ///
 /// When the cache `c` initially becomes invalid (i.e. it becomes
 /// potentially out of sync with the data source so that incremental updates
 /// would stop working), call `c.reset_on_next_modification()`; the next
 /// incremental update will then replace the entire contents of the cache,
 /// instead of incrementally augmenting it. Until that next modification,
 /// however, the stale contents of the cache will be made available.
-pub struct Cache<T> {
-    values: HashSet<T>,
+pub struct Cache<K, V> {
+    values: IndexMap<K, V>,
```

Comment:
This should be called `entries` or `inner`. It's confusing for it to be called "values" since "values" means something different for a map.